jws/jws.go

```diff
@@ -312,12 +365,37 @@ func verifyJSON(signed []byte, alg jwa.SignatureAlgorithm, key interface{}, dst
 		buf.WriteByte('.')
 		buf.WriteString(payload)
 
-		if err := verifier.Verify(buf.Bytes(), sig.signature, key); err == nil {
-			if dst != nil {
-				*dst = m
+		if !ctx.useJKU {
+			if hdr := sig.protected; hdr != nil && hdr.KeyID() != "" {
+				if jwkKey, ok := ctx.key.(jwk.Key); ok {
+					if jwkKey.KeyID() != hdr.KeyID() {
+						continue
+					}
+				}
+			}
+
+			verifier, err := NewVerifier(ctx.alg)
+			if err != nil {
+				return nil, errors.Wrap(err, "failed to create verifier")
+			}
+
+			if _, err := ctx.tryVerify(verifier, sig.protected, buf.Bytes(), sig.signature, m.payload); err == nil {
+				if ctx.dst != nil {
+					*(ctx.dst) = m
```

Comment:
not that I have any knowledge about how this works but I noticed this line and L393 are not covered by tests

Reply:
Thanks, addressed in 7a2ff0c